presto-main/src/main/java/com/facebook/presto/operator/MergeJoinSinkOperator.java

```diff
@@ -0,0 +1,159 @@
+        }
+
+        @Override
+        public MergeJoinSinkOperator createOperator(DriverContext driverContext)
+        {
+            checkState(!closed, "Factory is already closed");
+            OperatorContext operatorContext = driverContext.addOperatorContext(operatorId, planNodeId, MergeJoinSinkOperator.class.getSimpleName());
+            LocalMemoryContext memoryContext = operatorContext.aggregateUserMemoryContext().newLocalMemoryContext(MergeJoinSource.class.getSimpleName());
+            MergeJoinSource mergeJoinSource = mergeJoinSourceManager.getMergeJoinSource(driverContext.getLifespan(), memoryContext);
+            return new MergeJoinSinkOperator(operatorContext, mergeJoinSource);
+        }
+
+        @Override
+        public void noMoreOperators()
+        {
+            closed = true;
+        }
+
+        @Override
+        public OperatorFactory duplicate()
+        {
+            throw new UnsupportedOperationException("Callback sink can not be duplicated");
+        }
+    }
+
+    @VisibleForTesting
+    public enum State
+    {
+        CONSUMING_INPUT,
+        WAITING_FOR_CONSUMER,
+        CLOSED
+    }
+
+    private final OperatorContext operatorContext;
+    final MergeJoinSource mergeJoinSource;
+    private State state = State.CONSUMING_INPUT;
+
+    public MergeJoinSinkOperator(
+            OperatorContext operatorContext,
+            MergeJoinSource mergeJoinSource)
+    {
+        this.operatorContext = requireNonNull(operatorContext, "operatorContext is null");
+        this.mergeJoinSource = requireNonNull(mergeJoinSource, "mergeJoinSource is null");
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @VisibleForTesting
+    public State getState()
+    {
+        return state;
+    }
+
+    @Override
+    public ListenableFuture<?> isBlocked()
+    {
+        if (state == State.WAITING_FOR_CONSUMER) {
+            boolean canProduce = mergeJoinSource.getProducerFuture().isDone();
+            if (!canProduce) {
+                return mergeJoinSource.getProducerFuture();
+            }
+            state = State.CONSUMING_INPUT;
+        }
+        return NOT_BLOCKED;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return state == State.CONSUMING_INPUT;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        if (page == null || isFinished()) {
+            return;
+        }
+        checkState(state == State.CONSUMING_INPUT);
+
+        mergeJoinSource.addPage(page);
+        state = State.WAITING_FOR_CONSUMER;
+
+        operatorContext.recordOutput(page.getSizeInBytes(), page.getPositionCount());
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        return null;
+    }
+
+    @Override
+    public void finish()
+    {
+        if (state == State.CONSUMING_INPUT) {
+            mergeJoinSource.addPage(null);
+            state = State.CLOSED;
+        }
```

Comment:
1. move this logic to close(), and call close() directly
2. shouldn't we also handle situation WAITING_FOR_CONSUMER?

Reply:
So close() is called at the very end before destroying the operator. This logic is still a state change, and also signals the MergeJoinSource to finish, so makes more sense to keep it here.